### DISCUSSION TITLE
How to run openssh-server in LTSP clients

### DISCUSSION THREAD

**Discussion by jeremy-walker** [ltsp/ltsp#310] [Q&A] [answered]
I added the following to my LTSP server's `/etc/ltsp/ltsp.conf`:
```
POST_INIT_CP_KEYS="mv /etc/ltsp/ssh_host_* /etc/ssh/"
```
I then created a new set of ssh host keys in `/etc/ltsp`, followed by a `ltsp initrd`. When I check the client boot logs, I see the following message:
```
mv: cannot stat '/etc/ltsp/ssh_host_*': No such file or directory
```
After the client boots up, `ls -l /etc/ltsp` outputs the following:
```
total 16
-rw-r--r-- 1 root root  718 Jan  1  1970 group
-rw-rw---- 1 root sudo 1028 Jan  1  1970 ltsp.conf
-rw-r--r-- 1 root root 1711 Jan  1  1970 passwd
-rw-r--r-- 1 root root  852 Jan  1  1970 ssh_known_hosts
```

On the server however, `ls -l /srv/ltsp/ubuntu-2004/etc/ltsp` shows the following:
```
total 32
-rw------- 1 root root 1373 Sep  2 11:33 ssh_host_dsa_key
-rw-r--r-- 1 root root  598 Sep  2 11:33 ssh_host_dsa_key.pub
-rw------- 1 root root  505 Sep  2 11:33 ssh_host_ecdsa_key
-rw-r--r-- 1 root root  170 Sep  2 11:33 ssh_host_ecdsa_key.pub
-rw------- 1 root root  399 Sep  2 11:33 ssh_host_ed25519_key
-rw-r--r-- 1 root root   90 Sep  2 11:33 ssh_host_ed25519_key.pub
-rw------- 1 root root 2590 Sep  2 11:33 ssh_host_rsa_key
-rw-r--r-- 1 root root  562 Sep  2 11:33 ssh_host_rsa_key.pub
```
What am I doing wrong?

**Comment by alkisg** [comment]
I transferred this issue to https://github.com/ltsp/community which is for discussions.
While https://github.com/ltsp/ltsp is for bugs that can be solved in the LTSP code base.

What are you trying to do? To run openssh-server in the clients? Note that this is insecure, as you'll be publishing the ssh keys in all the local network. That's why epoptes-client is doing reverse connections, i.e. it relies on server-side secrets, not client-side secrets.

**Comment by jeremy-walker** [comment]
That's correct, I am running openssh-server on the clients. I had read a post somewhere on the LTSP forums that the workaround to exposing the server's ssh keys on the local network is to create another set of host keys, and then use the POST_INIT directive to copy those keys to `/etc/ssh` on the clients (I can't for the life of me find that post anymore). Doesn't this approach mitigate the risk somewhat?

**Comment by jeremy-walker** [comment]
Actually, I looked again and I found that post: https://github.com/ltsp/community/issues/87

**Comment by alkisg** [comment]
There are 3 ideas to discuss here:

1. Expose the server SSH keys (e.g. chrootless). This is the worst, security-wise. Of course this isn't enabled by default. To do that, one would put the following in ltsp.conf:

```shell
[server]
OMIT_FUNCTIONS="ssh_host_keys"
OMIT_IMAGE_EXCLUDES="etc/ssh/ssh_host_*"

[clients]
KEEP_SYSTEM_SERVICES="openssh-server"
```

2. Expose the chroot SSH keys, if one uses a chroot. This means that people in LAN will be able to get the private SSH keys of the clients. There are various discussions on the Internet about how bad it is if "the private SSH keys are exposed", no point in discussing "how bad it is" here. It's certainly better than exposing the server SSH keys, of course. The same ltsp.conf entries would be needed in this case as well. And, this idea is the same as "regenerating them on the server and copying them with POST_INIT_x".

3. Regenerate the keys in POST_INIT_x, while the client boots. This means that noone has the private SSH keys, but it also means that the server cannot trust the clients as they constantly change keys. This would only make some sense if the clients do not get rebooted frequently, e.g. weekly/monthly.

So I think that if you HAVE to use SSH in the clients, idea 2 is the better one.
People using chrootless that want to implement idea 2, can put customized SSH keys in /etc/ltsp/initrd/etc/ssh/*.

**Comment by jeremy-walker** [comment]
Thank you, that's a lot clearer now. 

So with option 2, is there a difference between the following two approaches?

1. Create a new set of keys on the server, then copy them to the client's chroot 
OR
2. Create a new set of keys on the server, then copy them to the server's `/etc/ltsp` and use a POST_INIT_x to copy them to the client's `/etc/ssh`.

**Comment by alkisg** [comment, marked as answer]
After re-reading the code, I remembered my design choices better, so let me post a fourth and better solution.

The SSH keys are provided by the additional initrd image (ltsp.img), in order for them to be common for all chroots and images. They are **not** provided by the x86_64.img squashfs image (that's why many commands were needed above to include them; there's no need for that). Also, the SSH keys are **not** included by default in the initrd for safety reasons, to avoid man-in-the-middle attacks.

To generate new keys and have them included in ltsp.img, commands similar to the following need to be executed on the server. I.e. the idea is to put the keys in /etc/ltsp/initrd/etc/ssh instead of the server or in a chroot:

```shell
sudo -i
# Create common SSH keys for all LTSP images:
mkdir -p /etc/ltsp/initrd/etc/ssh
for t in dsa ecdsa ed25519 rsa; do
    ssh-keygen -qf "/etc/ltsp/initrd/etc/ssh/ssh_host_${t}_key" -N '' -t "$t"
done
# Create and use a root SSH key
test -f /root/.ssh/id_rsa || ssh-keygen
mkdir -pm 700 /etc/ltsp/.ssh
install -m 700 /root/.ssh/id_rsa.pub /etc/ltsp/.ssh/authorized_keys
ltsp initrd
```

Additionally, the following is needed in ltsp.conf:

```shell
[clients]
KEEP_SYSTEM_SERVICES="ssh"
POST_INIT_MV_SSH="mv /etc/ltsp/initrd/etc/ssh/* /etc/ssh/ && mv /etc/ltsp/.ssh /root/"
```

**Reply by mherweg** [reply]
I think I  found a little mistake in the line:
install -m 700 /root/id_rsa.pub /etc/ltsp/.ssh/authorized_keys

better:
install -m 700 /root/.ssh/id_rsa.pub /etc/ltsp/.ssh/authorized_keys

**Reply by alkisg** [reply]
Thanks, fixed!

**Comment by jeremy-walker** [comment]
Thanks, that worked perfectly. 

The plan is to leave SSH access to the clients running for some time, just until the new setup stabilizes. Once everything is functioning smoothly, I intend to remove SSH access to the clients, since the only thing that we'll really need on the client side is a reboot whenever the chroot is updated. For that, I was thinking of running a cron job on the clients that polls for a "file" on one of the server's nfs shares, and when the "file" is found, it will reboot. Then all that'll be needed to reboot the clients is to run a `touch file` on the server.

**Comment by Vinpasso** [comment]
I'm trying to set this up for a compute cluster we have at our university. I placed the ssh host keys for the clients at the specified directory on the ltsp server:

```
pi@raspberrypi:~ $ ls /etc/ltsp/initrd/etc/ssh
ssh_host_dsa_key      ssh_host_ecdsa_key      ssh_host_ed25519_key      ssh_host_rsa_key
ssh_host_dsa_key.pub  ssh_host_ecdsa_key.pub  ssh_host_ed25519_key.pub  ssh_host_rsa_key.pub
```

I also added the ssh server as an exception to the clients in ltsp.conf:
```
KEEP_SYSTEM_SERVICES="ssh"
```

Then I ran the customary ltsp initrd:

```
pi@raspberrypi:~ $ sudo ltsp initrd
Generated ltsp.img:
-rw-r--r-- 1 root root 159232 Oct 22 19:22 /srv/tftp/ltsp/ltsp.img
```

When the clients start, the ssh server starts and I can connect but the connection is instantly closed back down:

```
pi@raspberrypi:~ $ ssh -v mai00
OpenSSH_7.9p1 Debian-10+deb10u2, OpenSSL 1.1.1d  10 Sep 2019
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: Applying options for *
debug1: Connecting to mai00 [192.168.67.100] port 22.
debug1: Connection established.
debug1: identity file /home/pi/.ssh/id_rsa type 0
debug1: identity file /home/pi/.ssh/id_rsa-cert type -1
debug1: identity file /home/pi/.ssh/id_dsa type -1
debug1: identity file /home/pi/.ssh/id_dsa-cert type -1
debug1: identity file /home/pi/.ssh/id_ecdsa type -1
debug1: identity file /home/pi/.ssh/id_ecdsa-cert type -1
debug1: identity file /home/pi/.ssh/id_ed25519 type -1
debug1: identity file /home/pi/.ssh/id_ed25519-cert type -1
debug1: identity file /home/pi/.ssh/id_xmss type -1
debug1: identity file /home/pi/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_7.9p1 Debian-10+deb10u2
debug1: Remote protocol version 2.0, remote software version OpenSSH_7.9p1 Debian-10+deb10u2
debug1: match: OpenSSH_7.9p1 Debian-10+deb10u2 pat OpenSSH* compat 0x04000000
debug1: Authenticating to mai00:22 as 'pi'
debug1: SSH2_MSG_KEXINIT sent
Connection reset by 192.168.67.100 port 22
```

Turns out that the ssh server on the client is missing its host certificates (I can't show the logs as I cannot easily copy them out of the client without connecting by ssh). I checked, and it looks like the host keys were located under `/etc/ltsp/initrd/etc/ssh` on the clients rather than on `/etc/ssh`. Is there something I missed or did wrong?

I searched through the source but I didn't find a place where the `/etc/ltsp/initrd` directory is copied to the root of the initrd image, only one that copies it to the `/etc/ltsp` directory on the initrd image. Is that intended?

Thanks for your help and have a nice evening.

**Comment by alkisg** [comment]
Hi Vinpasso, I believe you're right. I see comments in the code mentioning that /etc/ltsp/initrd is copied to the rootfs, but I think I forgot to actually implement that.
I'll rework it in the future, for now you could just mv the files with an ltsp.conf parameter:

```shell
POST_INIT_MV_SSH="mv /etc/ltsp/initrd/etc/ssh/* /etc/ssh/"
```

**Comment by alkisg** [comment]
I updated the [solution](https://github.com/ltsp/ltsp/discussions/310#discussioncomment-101549) with better instructions.

**Reply by Vinpasso** [reply]
Looks good to me. Adding the `POST_INIT_MV_SSH` script made it work and I haven't had any problems since then.
